### DISCUSSION TITLE
Custom fields as columns in collection table

### DISCUSSION THREAD

**Discussion by city41** [avitorio/outstatic#152] [Ideas]
I would like it if custom fields showed up in a collection's table.

Something like when creating a custom field, adding a boolean for "display in table"

If that boolean is checked, then when looking at a collection table, that field becomes a column.

I am happy to take a stab at this and create a PR if there is interest in this.

**Comment by avitorio** [comment]
Hey, this is a great suggestion. What do you think, instead of making this an option from the "Edit Collection" page we do something like this:

![image](https://github.com/avitorio/outstatic/assets/1417109/ddb004c0-fb18-40ba-8d26-5271532a0a05)

Feel free to create a PR and take a shot!

**Reply by avitorio** [reply]
Btw, these could work as checkmarks.
Not a requirement but it would be cool to use Radix-ui primitives or shadcn/ui for these.

**Reply by city41** [reply]
I like that because I would also ideally remove some of the default columns. 

I'll take a stab at this, but I probably won't really get a chance to really look into it for at least a week or two.

**Reply by avitorio** [reply]
No problem! Whenever you can, all help is welcome :)

**Reply by city41** [reply]
I was thinking I'd first store the column choice in local state. So the user would have to rechoose their columns each time.

Then from there some kind of persistence. Localstorage would be simple to do and pretty effective. But possibly the more proper place is a file in the repo like schema.json is. What do you think?

**Reply by avitorio** [reply]
I'm thinking that the place to store this info is in schema.json
Every collection, with custom fieldss, creates a schema.json file.
Perhaps adding a `list` field in the schema.json that holds an array with the name/order of visible fields?

```
{
  "title": "houses",
  "type": "object",
  "list": ["bathrooms", "bedrooms"],
  "properties": {
    "bedrooms": {
      "required": false,
      "description": "Number of bedrooms",
      "fieldType": "Number",
      "dataType": "number",
      "title": "Bedrooms"
    },
    "bathrooms": {
      "required": false,
      "description": "Number of bathrooms",
      "fieldType": "Number",
      "dataType": "number",
      "title": "Bedrooms"
    },
    "garages": {
      "required": false,
      "description": "Number of garages",
      "fieldType": "Number",
      "dataType": "number",
      "title": "Bedrooms"
    }
  }
}
```

**Reply by avitorio** [reply]
Remember that if a collection is deleted we should also remove it from the list.

**Reply by avitorio** [reply]
What do you think?

**Reply by city41** [reply]
Yup, makes sense. This would decide columns for all users but that is probably fine. If not could key it by user id. I would guess a lot of outstatic usages only have one user (I have two)

**Reply by avitorio** [reply]
Hmmm... I don't like the idea of keying by user id just because I don't think it's a good idea to store user data in git.
Perhapss a mix of default and localhost could work in the future.
Set default on the /collections/[collection] path and localhost if the user makes the change in the list view...
Maybe a v2.0 thing.

**Reply by avitorio** [reply]
I created an issue for it here: https://github.com/avitorio/outstatic/issues/153

**Reply by city41** [reply]
Sorry to say we've decided a different route for our CMS, so I won't be doing this work after all.

**Reply by avitorio** [reply]
No problem, thanks for suggesting the feature regardless. If you don't mind me asking, what will you use and what did we miss?

**Reply by city41** [reply]
I whipped up a Google doc as very primitive and crappy cms. I did this before finding outstatic. We just decided to stick with it for now, but may revisit outstatic in the future.

**Comment by avitorio** [comment]
Hey @city41 ,
Just wanted to let you know we've added the `Custom fields as columns in collection table` feature! :)
